### PR TITLE
filtering the xref of '/' based on authorization

### LINKED ISSUE #1610: can see unauthorized projects in /source/xref
The `/source/xref` listing contains projects the user is non authorized to.

### PR DESCRIPTION
**Pull request by tulinkry**
fixes #1610

### CODE REVIEW

File: src/org/opensolaris/opengrok/web/PageConfig.java [hunk 1]
```diff
@@ -391,8 +392,22 @@ public String canProcess() {
...
-                        = Collections.unmodifiableList(Arrays.asList(files));
+                List<String> list = Arrays.asList(files);
+                if (getPath().isEmpty()) {
+                    /**
+                     * This denotes the source root directory, we need to filter
+                     * projects which aren't allowed by the authorization. E. g.
+                     * remove all which aren't among the filtered set of
+                     * projects.
+                     */
+                    list = new ArrayList<>(list);
```
Comment (vladak) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: I'd rather have the `list` variable named differently than the original, will make it easier to debug.

File: src/org/opensolaris/opengrok/web/PageConfig.java [hunk 2]
```diff
@@ -391,8 +392,22 @@ public String canProcess() {
...
+                    list.removeIf((t) -> {
+                        return !getProjectHelper().getAllProjects().stream().anyMatch((p) -> {
```
Comment (vladak) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: what happens if projects are disabled ?

File: src/org/opensolaris/opengrok/web/PageConfig.java [hunk 3]
```diff
@@ -391,8 +392,22 @@ public String canProcess() {
                 dirFileList = Collections.emptyList();
             } else {
                 Arrays.sort(files, String.CASE_INSENSITIVE_ORDER);
-                dirFileList
-                        = Collections.unmodifiableList(Arrays.asList(files));
+                List<String> list = Arrays.asList(files);
+                if (getPath().isEmpty()) {
+                    /**
+                     * This denotes the source root directory, we need to filter
+                     * projects which aren't allowed by the authorization. E. g.
```
Comment (vladak) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: adjust the comment explaining what method called below actually performs the authorization check

File: src/org/opensolaris/opengrok/web/PageConfig.java [hunk 4]
```diff
@@ -391,8 +392,22 @@ public String canProcess() {
                 dirFileList = Collections.emptyList();
             } else {
                 Arrays.sort(files, String.CASE_INSENSITIVE_ORDER);
-                dirFileList
-                        = Collections.unmodifiableList(Arrays.asList(files));
+                List<String> list = Arrays.asList(files);
+                if (getPath().isEmpty()) {
+                    /**
+                     * This denotes the source root directory, we need to filter
```
Comment (vladak) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: pls explain in the comment why this only has to be done for the root dir

File: src/org/opensolaris/opengrok/web/PageConfig.java [hunk 5]
```diff
@@ -377,7 +377,8 @@ public String canProcess() {
      *
      * @return an empty list, if the resource does not exist, is not a directory
      * or an error occurred when reading it, otherwise a list of filenames in
-     * that directory, sorted alphabetically
+     * that directory, sorted alphabetically and filtered based on authorization
```
Comment (vladak) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: not precise I think as the authorization is performed only for root dir

File: src/org/opensolaris/opengrok/web/PageConfig.java (hunk 2 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: that's a good question also for the whole auth framework

File: src/org/opensolaris/opengrok/web/PageConfig.java (hunk 1 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: I changed that

File: src/org/opensolaris/opengrok/web/PageConfig.java (hunk 2 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: fixed here

File: src/org/opensolaris/opengrok/web/PageConfig.java (hunk 3 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: done

File: src/org/opensolaris/opengrok/web/PageConfig.java (hunk 4 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: done

File: src/org/opensolaris/opengrok/web/PageConfig.java (hunk 5 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: changed

File: src/org/opensolaris/opengrok/web/PageConfig.java [hunk 6]
```diff
@@ -378,6 +378,11 @@ public String canProcess() {
      * @return an empty list, if the resource does not exist, is not a directory
      * or an error occurred when reading it, otherwise a list of filenames in
      * that directory, sorted alphabetically
+     *
+     * <p>
+     * For the root directory (/xref/) an authorization is performed for each of
+     * the project in case that projects are used in OpenGrok.</p>
```
Comment (vladak) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: 'each project' or 'each of the projects'

Comment (vladak) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: 'in OpenGrok' is redundant

**Comment by vladak** [comment]
Is this possible to unit test this ?

File: src/org/opensolaris/opengrok/web/PageConfig.java (hunk 6 above)
Comment (tulinkry) [inline src/org/opensolaris/opengrok/web/PageConfig.java]: fixed

**Comment by tulinkry** [comment]
I added a simple test case for this

**Pull request merged**